storey/targets.py

```
@@ -888,48 +891,100 @@ def _init(self):
+        return cls.UnsupportedTDEngineTypeError(f"Unsupported value type {type(value)} in {cls.__name__}")
+
+    @classmethod
+    def _value_to_tag(cls, value):
+        import taosws
+
+        if isinstance(value, bool):
+            return taosws.bool_to_tag(value)
+        if isinstance(value, int):
+            return taosws.int_to_tag(value)
+        if isinstance(value, float):
+            return taosws.float_to_tag(value)
+        if isinstance(value, str):
+            return taosws.varchar_to_tag(value)
+        raise cls._get_unsupported_error(value)
+
+    def _get_tags_from_event(self, event: dict) -> list:
+        tags = []
+        for tag_name in self._columns[: self._number_of_tags]:
+            tags.append(self._value_to_tag(event.get(tag_name)))
+        return tags
+
+    @classmethod
+    def _raw_value_to_column_fun(cls, value) -> Callable:
+        import taosws
+
+        if isinstance(value, datetime.datetime):
+            return taosws.millis_timestamps_to_column
+        if isinstance(value, bool):
+            return taosws.bools_to_column
+        if isinstance(value, int):
+            return taosws.ints_to_column
+        if isinstance(value, float):
+            return taosws.floats_to_column
+        if isinstance(value, str):
+            return taosws.nchar_to_column
+        raise cls._get_unsupported_error(value)
+
+    @classmethod
+    def _raw_value_to_value(cls, value):
+        if isinstance(value, datetime.datetime):
+            return int(value.timestamp() * 1000)
+        return value
+
+    def _get_batch_values(self, batch: list[dict]) -> list:
+        values = [[] for _ in range(self._number_of_values)]
+        to_column_funs = []
+        val_names = self._columns[self._number_of_tags :]
+        for event_ind, event in enumerate(batch):
+            for val_ind, val_name in enumerate(val_names):
+                raw_value = event.get(val_name)
```

Need to raise an error if missing (or is that ok in some cases – i.e. nullable columns?). As it's written, it will raise a
```UnsupportedTDEngineTypeError("Unsupported value type NoneType in TDEngineTarget")```

---

I'll handle the type resolution with the table's schema - obtained through a `DESCRIBE` query.

---

Done